### DISCUSSION TITLE
Can't get background/quit state push notification events

### DISCUSSION THREAD

**Discussion by aevaldas** [invertase/react-native-firebase#4235] [Q&A] [answered]
I can see a push notification when my app is on background or quit state, but I can't get console output.

Handling background state messages like that (index.js):
```
messaging().setBackgroundMessageHandler(async (remoteMessage) => {
  console.log(
    'Notification received on background or quit state:',
    remoteMessage,
  );
});
```

How should I debug it?

**Comment by mikehardy** [comment]
Please provide versions in use, and the exact JSON sent from the firebase admin sdk (minus device tokens or API tokens of course), plus a full App.js to reproduce

You don't even mention what operating system you are using :sweat_smile: - how would you answer this question if a colleague came to you, asking it?

As a guess though, you have not set the handler prior to registering the app root component?

**Reply by aevaldas** [reply]
Sorry for lack of information and thank you for warm answer 😅

I've tried to make a clear react native setup with: @react-native-firebase/app, @react-native-firebase/messaging, @react-native-firebase/analytics.

I have prepared a public repo: https://github.com/aevaldas/RNFirebaseTestEA
I hadn't not included the development team on Xcode here.

package.json:
```
    "@react-native-firebase/analytics": "^7.6.1",
    "@react-native-firebase/app": "^8.4.1",
    "@react-native-firebase/messaging": "^7.8.4",
    "react": "16.13.1",
    "react-native": "0.63.2"
```

#### FCM request

`https://fcm.googleapis.com/fcm/send`

headers:
```
Authorization: key=...
```
body:
```
{
    "registration_ids": [
        "..."
    ],
    "content_available": true,
    "priority": "high",
    "notification": {
        "title": "Title",
        "body": "Description"
    },
    "data": {
        "test": "1"
    }
}
```

Problems: 

- On first load Android does not receive push notifications, but once the app is reloaded, newly sent push notifications is delivered on foreground, background and quit state.

- iOS does not receive any push notifications, despite the fact that it receives the FCM token.

--

> As a guess though, you have not set the handler prior to registering the app root component?

I guess I missed that part from docs, could you reference me there?

**Comment by mikehardy** [comment, marked as answer]
If you include notification payload on iOS, when you are not in the foreground, you asking the SDKs to handle the JSON when it comes in. You won't get control (in the handler) until the user interacts with the notification

https://github.com/aevaldas/RNFirebaseTestEA/blob/9482735df19da5305473e4afa14d92b6369540ed/index.js#L10-L12

Looks correct but if you want the background/quit state stuff you need to go data only I think

**Reply by mikehardy** [reply]
Also run it on a real device in release mode while plugged in and watching logs from Console.app or similar, there are lots of reasons you might be throttled on device and it rejects the message
